torch_spyre/csrc/spyre_mem.cpp

```diff
@@ -44,6 +48,411 @@
 
 namespace spyre {
 
+struct DMAParameters {
+  const int64_t size_bytes;  // bytes to transfer
+  const off64_t src_offset;
+  const off64_t dst_offset;
+};
+auto get_device_shape(c10::IntArrayRef sizes) -> std::vector<int64_t> {
+  /* Given the CPU Tensor's shape, return the shape of the tensor on Spyre
+   */
+  auto cpu_shape = sizes.vec();
+  std::vector<int64_t> dev_shape;
+  auto stick_size = 64;
+  auto requires_padding = (cpu_shape.back() % stick_size != 0);
+
+  for (int i = 0; i < cpu_shape.size(); i++) {
+    if (i == 0 && cpu_shape.size() == 1) {
+      dev_shape.push_back(stick_size);
+      if (requires_padding) {
+        dev_shape.push_back(1);
+      } else {
+        dev_shape.push_back(cpu_shape.back() / stick_size);
+      }
+    } else if (i == 0) {
+      dev_shape.push_back(stick_size);
+    } else if (i == sizes.size() - 1) {
+      if (cpu_shape.size() <= 2) dev_shape.push_back(cpu_shape[i - 1]);
+      if (requires_padding) {
+        auto padded_shape =
+            cpu_shape.back() + (stick_size - (cpu_shape.back() % stick_size));
+        dev_shape.push_back(padded_shape / stick_size);
+      } else {
+        dev_shape.push_back(cpu_shape.back() / stick_size);
+      }
+      if (cpu_shape.size() > 2) dev_shape.push_back(cpu_shape[i - 1]);
+    } else {
+      dev_shape.push_back(cpu_shape[i - 1]);
+    }
+  }
+  std::reverse(dev_shape.begin(), dev_shape.end());
+  return dev_shape;
+}
+auto get_device_shape(const at::Tensor* tensor) -> std::vector<int64_t> {
+  /* Given the CPU Tensor, return the shape of the equivalent tensor on
+   * Spyre
+   */
+  const c10::IntArrayRef& sizes = tensor->sizes();
+  return get_device_shape(sizes);
+}
+auto get_device_layout(c10::IntArrayRef sizes, c10::IntArrayRef strides)
```

Comment:
strides are ignored here, right? put it in a comment and maybe make a TODO to support strided tensors later